src/node_sqlite.cc

```diff
@@ -773,15 +788,25 @@ void DatabaseSync::ApplyChangeset(const FunctionCallbackInfo<Value>& args) {
         options->Get(env->context(), env->onconflict_string()).ToLocalChecked();
 
     if (!conflictValue->IsUndefined()) {
-      if (!conflictValue->IsNumber()) {
+      if (!conflictValue->IsFunction()) {
         THROW_ERR_INVALID_ARG_TYPE(
             env->isolate(),
-            "The \"options.onConflict\" argument must be a number.");
+            "The \"options.onConflict\" argument must be a function.");
         return;
       }
-
-      int conflictInt = conflictValue->Int32Value(env->context()).FromJust();
-      conflictCallback = [conflictInt]() -> int { return conflictInt; };
+      Local<Function> conflictFunc = conflictValue.As<Function>();
+      conflictCallback = [env, conflictFunc](int conflictType) -> int {
+        Local<Value> argv[] = {Integer::New(env->isolate(), conflictType)};
+        TryCatch try_catch(env->isolate());
+        Local<Value> result =
+            conflictFunc->Call(env->context(), Null(env->isolate()), 1, argv)
+                .FromMaybe(Local<Value>());
+        if (try_catch.HasCaught()) {
+          try_catch.ReThrow();
+          return SQLITE_CHANGESET_ABORT;
+        }
+        return result->Int32Value(env->context()).FromJust();
```

Comment:
Returning a non-number from the callback essentially results in undefined behaviour, as the value is coerced to int by the call to `Value::Int32Value()` before it's validated downstream by the sqlite library.

This is a tad inconsistent, as returning an invalid number _will_ raise an error downstream, but returning eg. an object _won't_ (as it'll just be silently coerced to `0`). To keep things consistent, it would probably be worth passing a deliberately invalid value back to sqlite in the case where `Value::IsInt32()` is false.

Reply:
Good catch, thank you.

Reply:
Please see https://github.com/nodejs/node/pull/56352/commits/5d2fc4e38fe37732c535e687cc4534aa88a83ad9